File: Stack/239.sliding-window-maximum.go
package leetcode

// @leet start
type s239 struct {
	ans    []int
	window []int
}

func (s *s239) push(v int) {
	for len(s.window) > 0 && v > s.window[len(s.window)-1] {
		s.window = s.window[:len(s.window)-1]
	}
	// push only if v <= s.lastItem
	s.window = append(s.window, v)
}

func (s *s239) pop(v int) {
	if len(s.window) > 0 && s.window[0] == v {
		s.window = s.window[1:]
	}
}

func (s *s239) update() {
	s.ans = append(s.ans, s.window[0])
}

func maxSlidingWindow(nums []int, k int) []int {
	s := s239{}
	// fill window for first k elmement
	for i, v := range nums {
		if i == k {
			break
		}
		s.push(v)
	}
	s.update()
	// loop window
	for i := k; i < len(nums); i++ {
		s.push(nums[i])
		s.pop(nums[i-k])
		s.update()
	}
	return s.ans
}

// @leet end

File: Stack/239.sliding-window-maximum.py
from collections import deque
from typing import List

def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        ret = []
        q = deque()

        def push(v):
            while q and q[-1] < v:
                q.pop()
            q.append(v)

        def pop(v):
            if q[0] == v:
                q.popleft()

        def update():
            ret.append(q[0])

        for i in range(k):
            push(nums[i])
        update()
        for i in range(k, len(nums)):
            push(nums[i])
            pop(nums[i - k])
            update()

        # for i in range(len(nums)):
        #     if i < k - 1:
        #         push(nums[i])
        #     else:
        #         push(nums[i])
        #         update()
        #         pop(nums[i - k + 1])
        return ret
# @leet start
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        res, stack = [], deque()

        def push(v):
            while len(stack) > 0 and v > stack[-1]:
                stack.pop()
            stack.append(v)

        def pop(v):
            if len(stack) > 0 and v == stack[0]:
                stack.popleft()

        def update():
            res.append(stack[0])

        for i, v in enumerate(nums):
            if i == k:
                break
            push(v)
        update()
        for i in range(k, len(nums)):
            push(nums[i])
            pop(nums[i - k])
            update()
        return res


# @leet end

